### ISSUE TITLE
Calls two files duplicates when they definitely aren't (different sizes, even)

### ISSUE THREAD

**Issue by b3taman** [kornelski/dupe-krill#12]
Hi Kornel,

You may want to check how this happened.

**TL;DR:** I'm glad I did a dry run, as otherwise I'd have lost REALLY important encrypted information.

I thought about it a bit, and decided to give what I hope will be useful input. After all, fixing it would yield something really useful to me as well ;-)

---

Having decades of backups spread across everything from tape and floppies to SSD and cloud, I really need something like this, so decided to give it a whirl. Unfortunately the first dummy run I did showed that running it would have been a disaster for me.

I installed using your .deb file.

Here is the output:

```
USER@HOST:~$ dupe-krill -d .
Dry run. No files will be changed.
Found dupe /USER/{e => .}/crpt.tar.gpg
Found dupe /USER/{e => .}/2fa.tar.gpg
...
USER@HOST:~$ diff crpt.tar.gpg 2fa.tar.gpg 
Binary files crpt.tar.gpg and 2fa.tar.gpg differ
USER@HOST:~$ ls -l crpt.tar.gpg 2fa.tar.gpg 
-rw-r--r-- 1 USER USER  436708 Dec  4  2019 2fa.tar.gpg
-rw-r--r-- 1 USER USER 8216343 Dec  4  2019 crpt.tar.gpg
USER@HOST:~$ 
```

I had a quick glance at some of the code. I've never written (or read) Rust before, so there were a few weird / interesting things, but a basic question: are you first testing for reasons to reject files from further processing? You should do that before anything else.

Based on an empirical test I did (shown below the algorithm), a high level, off-the-cuff algorithm might be as follows (depending on -v, -vv, etc., inform the user as appropriate):

**Initialisation**

1. Memory & disc space checks, bombing out with explanation if required, or warning and asking for manual intervention unless called with, say, -f (for force: I don't care if every operation involves swap space; I'm going on holiday and want it to work while I'm gone.).

2. Build an in-memory tree of the FS level information of the search domain, i.e. file names, sizes, etc. See my empirical results below.

3. Hard links only work within a single file system, so check if there is more than one FS in any of the search paths and prune those from the tree. Continue if anything remains.

4. Traverse the tree, classifying and pruning further, if possible (removing soft links, for example). Also remove certain types of cache files or other speed-up file types as per my "Another thought" below.

**Iteration**

This is really high level but gives the idea. Since we did a lot of useful things during initialisation, there is not a lot to do at each iteration. I would test recursive vs. non-recursive implementations, BTW.

1. If file sizes are different, then ignore the pair under consideration.

2. If all other reasons to ignore the pair do not apply, then compare and de-duplicate if all pass.

Another thought: From horrible experience of Mac users polluting my drives, MacOS uses things like .DS_Store and other artefacts. Some of those may start off identical, be hard-linked by your program, then when one location changes and the OS changes the file, it messes up the other areas which also use that file, as the contents now no longer reflect the reality of their (now different) environments. (See https://en.wikipedia.org/wiki/.DS_Store for a bit on this). This applies to all manner of caches, making automated use quite a challenge.

My empirical test was quite simple and showed that the above approach should be fine (it was done on an Intel Core i7-3520M, which is quite old, being 3rd generation):

```
root@HOST:/ # time find -type l | wc -l
301708

real	2m56.858s
user	0m1.541s
sys	0m6.398s
root@HOST:/ # time find | wc -l
1611379

real	0m2.275s
user	0m0.861s
HOSTsys	0m1.555s
root@HOST:/ # time find -type d | wc -l
99209

real	0m1.995s
user	0m0.624s
sys	0m1.378s
root@HOST:/ # time find -type f | wc -l
1210085

real	0m2.163s
user	0m0.788s
sys	0m1.505s
root@HOST:/ # free
              total        used        free      shared  buff/cache   available
Mem:        8073452     2584000     2481952      466704     3007500     4722640
Swap:       8650748           0     8650748
root@HOST:/ #
```

Well, there you go! I hope this helps, as running the code showed me that it at least did not ignore the two .gpg files based on size. As to why it thought they are similar, I have no idea.

PS: Those weren't the only ones. The small test I did returned the following in its JSON output:

```  
  "stats": {
    "added": 145,
    "skipped": 326,
    "dupes": 8,
    "hardlinks": 0
  }
```

All the identified duplicates were wrong, but one case should suffice.

8-)
John (b3ta)

**Comment by kornelski** [comment]
Thanks for the report. That is very concerning indeed.

Which version did you use?

**Comment by b3taman** [comment]
So much fro a complete report ;-)

It was v1.4.4, which I downloaded a few hours ago.

**Comment by kornelski** [comment]
I'm not sure how to investigate this, because of course I haven't seen any failure like that *on my machine*.

Do these files have identical prefix? 

Are you able to run Rust tests? 

1. Get Rust/Cargo from https://rustup.rs
2. Clone the repo
3. Run `cargo test`

**Comment by b3taman** [comment]
**TL;DR:** It is working, though the documentation could be made a bit clearer and I should assume less :-)
///
I created a few test cases and it turns out that the issue was the terseness of the output combined with the terseness of my directory names ("0" and "e"). Maybe explain how you format your output, since I was confused by that and my assumptions.

I found it interesting during testing that once a duplicate was found, it compared all the others to the initial _duplicate_ name as opposed to the actual first one (both in terms of time of creation and naming as well as alphabetically). Is that your algorithm, or an artefact of how Rust handles its data structures?

### Thanks for 
### I have now run it over ### files and 

### As a matter of interest, I'd appreciate your thoughts on the point I made about the .DS-Store and similar files, as that could be a deal breaker for many people. It works in backups (as long as the restore restores to individual files), but I see danger doing that on live data. It might be worth

**Comment by b3taman** [comment]
[Second attempt. I *didn't* close it — a software KVM wreaked havoc when I thought I was on another computer after it crashed under me.]

**TL;DR:** It is working, though the documentation could be made a bit clearer and I should assume less :-) Thus: some egg on my face.

That said, I have a few concerns, but will mail you directly after running a surface test and doing a file system integrity check.

///

Two things:

1. I created a few test cases and it turns out that the issue was the terseness of the output combined with the terseness of my directory names ("0" and "e"). Maybe explain how you format your output, since I was confused by that and my assumptions.

1. I found it interesting during testing that once a duplicate was found, it reported the others relative to the initial _duplicate_ name as opposed to the first name listed. (First in terms of time of creation/modification, alphabetically, and as reported in your output). Is that your algorithm, or an artefact of how Rust handles its data structures?

**Comment by kornelski** [comment]
Ok, good to know it wasn't an error!

1. I've added a note about the path syntax in the readme.

2. Merging has several trade-offs.
   * Directories are scanned in inode order (lower inode number first). This is a small optimization that tries to help filesystems scan things more sequentially for them. 
   * Files are hardlinked into the file that already has the most hardlinks, so that merging of multiple groups of hardlinks does least amount of I/O.
